### DISCUSSION TITLE
Read-only document display

### DISCUSSION THREAD

**Discussion by matthew-carroll** [Flutter-Bounty-Hunters/super_editor#303] [Milestones]
Super Editor should support the display of documents in a read-only manner.

Use-cases:

- Display a blog post, news article, etc
- Display an editable document to a read-only participant (like giving someone "view" access to a Google Doc)
- Display an email to a recipient. The email was originally editable by the author, but is read-only for the recipient
- Easy switch between a read-only and editable document, e.g. this github editor which allows switching between preview and editing or on on mobile to go into editing on single tap, while dragging should scroll (and not select text)

Funding: Clearful

**Comment by osaxma** [comment]
Hi @matthew-carroll, 

I’ve done an experiment before for a simple read-only view that may be of interest. 

I pulled the code out in a separate branch with a demo that can be found here: [read_only_experiment](https://github.com/osaxma/super_editor/tree/read_only_experiment) :: [diff](https://github.com/superlistapp/super_editor/compare/main...osaxma:read_only_experiment).

In short, I created `demo_read_only.dart` where `SuperEditor` simply hides the caret and limits the keyboard actions to the following:  select all, copy selection, and expand selection with arrows. This is achieved by exposing `showCaret` flag and customizing the keyboard actions. I believe this covers the first two use cases for a simple viewer. Any additional features such as highlighting text or adding comments can be achieved in a similar manner to the styling toolbar of the example editor. 

As for the third use case, I've found that component builders need to be aware of the editing mode somehow. For instance, in an edit mode, a custom image component may show an `X` mark at a corner to delete the image whereas in a read-only mode, it'll show a magnifying glass instead to enlarge the image. Same applies to the email field if it was a component. To achieve this, I've added a read-only flag to SuperEditor that gets passed to component builders through `ComponentContext`.  

While this is not implemented in the experiment, the read-only flag could also be useful for the following:
- Determine when arrow key presses are supposed to scroll the view instead of moving the caret.
- Prevent the IME keyboard in mobile from showing up in a read-only state. 

Additional consideration:
- This only works with `MutableDocument` since it's the only accepted type by `DocumentEditor`.

**Comment by matthew-carroll** [comment]
This milestone is now complete with the addition of `SuperReader`: https://github.com/superlistapp/super_editor/pull/789

If we want to extend the read-only capabilities, we'll file another milestone.
